scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java

```diff
@@ -27,89 +29,88 @@
+        if (rootLocationString != null) {
+            return rootLocationString;
+        } else {
+            return "";
+        }
+    }
+
     @Override
-    public void addDatasetDirectories(String id, String username) throws DatasetStorageException {
-        // First, the correct directory must be created
-        Path path = createPath(username, id);
+    public boolean addDirectories(String username, String type, String id, String version) throws
+            FileSystemStorageException {
+        if (username == null) {
+            return false;
+        }
+        if ((type == null || id == null || version == null) &&
+                (type != null || id != null || version != null)) {
+            return false;
+        }
+        Path path = createPath(username, type, id, version);
         File directory = path.toFile();
         // if the Path was created successfully, create the directory
         try {
             if (!directory.exists()) {
                 Files.createDirectories(path);
             }
         } catch (IOException e) {
-            throw new DatasetStorageException("IOException occurred when attempting to add the Dataset", e);
+            throw new FileSystemStorageException("IOException occurred when attempting to add the Dataset", e);
         } catch (UnsupportedOperationException e) {
-            throw new DatasetStorageException("UnsupportedOperationException occurred when attempting to add the Dataset", e);
+            throw new FileSystemStorageException("UnsupportedOperationException occurred when attempting to add the Dataset", e);
         } catch (SecurityException e) {
-            throw new DatasetStorageException("SecurityException occurred when attempting to add the Dataset", e);
+            throw new FileSystemStorageException("SecurityException occurred when attempting to add the Dataset", e);
         }
+        return true;
     }
 
-    /**
-     * A helper method to create a Path given a username and dataset name.
-     *
-     * @param username  The username of the owner of the dataset.
-     * @param datasetId The id of the dataset.
-     * @return Returns a Path.
-     */
     @Override
-    public Path createPath(String username, String datasetId) {
+    public Path createPath(String username, String type, String id, String version) throws
+            FileSystemStorageException {
         Path path;
         try {
-            if (datasetId != null) {
-                path = Paths.get(this.rootLocationString, username, datasetId);
+            if (username != null && (type == null && id == null && version == null)) {
+                path = Paths.get(getRootLocationString(), username);
+            } else if (username != null && (type != null && id != null && version != null)) {
+                path = Paths.get(getRootLocationString(), username, type, id, version);
+            } else if (username != null && (type != null && id != null)) {
```

Comment:
you don't need `()` around `&&` statements; `true && (false && true)` is the same as `true && false && true`.